RESTORE can use spatially varying sigma

samcoveney opened the pull request:
I have adjusted the function `restore_fit_tensor` to be able to accept a spatially varying sigma, rather than only a sigma that varies for each image (or a constant sigma). In this way, the noise estimate from functions such as `pca_noise_estimate` can be used (I have not allowed to have a sigma that varies across both space and image, because I can't see the use case for this).

I have left notes in this file, with questions and thoughts, because I have made a few other adjustments (some of which could be applied to `nlls_fit_tensor` as well).

Very much a work in progress, but fundamentally the changes are fairly simple. I just need some feedback.

edit: also addresses #2574 for RESTORE

arokem commented:
Hi @samcoveney : thanks for working on this! Looks great so far, and I will try to go through and address your inline questions/comments. My first general comment is that it would be great to add some testing for the new functionality. A particular point I noticed is that things might fail in weird ways if the user provides a scalar as a numpy array for sigma (i.e., `sigma = np.array([a])`), but I might be missing something.

arokem commented on dipy/reconst/dti.py:
```diff
@@ -1801,90 +1850,102 @@ def restore_fit_tensor(design_matrix, data, sigma=None, jac=True,
+                params[vox, :3] = evals
+                params[vox, 3:12] = evecs.ravel()
+            else:
+                # NOTE: how should we record the success of RESTORE? Could have
```
If RESTORE fails because the sigma was off (as in #2574), raising a warning might be appropriate.

samcoveney replied:
Warnings are raised for failed fitting, and option `fail_is_nan` should will indicate when RESTORE has failed

arokem reviewed (commented):
OK - I looked through the code and added some comments/questions.

samcoveney commented on dipy/reconst/dti.py:
```diff
@@ -1378,12 +1394,28 @@ def wls_fit_tensor(design_matrix, data, return_S0_hat=False):
+        fit_result = np.einsum('...ij,...j',
+                               pinv(design_matrix * w[..., None]),
+                               w * log_s)
+    else: # faster - this is odd, since pinv is supposed to be faster
```
To benchmark the speed, I just put some 100 iteration loops around the code and timed it, but of course this is highly specific on the data I was testing on...

The documentation for pinv function in dipy.utils.arrfuncs suggests that this should be used when numpy is below a particular version, but np.linalg.pinv is now vectorized. Perhaps np.linalg.pinv should be preferred based on the numpy version number?  (It is bound to be faster than the implementation in pinv, which my simple tests would suggest)

samcoveney commented on dipy/reconst/dti.py:
```diff
@@ -1378,12 +1394,28 @@ def wls_fit_tensor(design_matrix, data, return_S0_hat=False):
+    if False:  # use _ols_fit_matrix - slow
+        ols_fit = _ols_fit_matrix(design_matrix)  # ols_fit is "hat matrix", y' = ols_fit @ y
+        w = np.exp(np.einsum('...ij,...j', ols_fit, log_s))  # predicted S
+    else:  # use same method as ols_fit_matrix - fast
```
There are two functions, `ols_fit_matrix` and `_ols_fit_matrix`. The latter returns the 'hat matrix', which when multiplied by data y-values, returns predicted 'y-values'. The only purpose for `_ols_fit_matrix` according to docs is "Mainly used to calculate WLS weights". Given that np.linalg.pinv uses SVD anyway, and would appear to be slower method of getting to the weights, I would might suggest actually removingn `_ols_fit_matrix` entirely, since it is confusing and would appear to be used nowhere else at all

samcoveney commented on dipy/reconst/dti.py:
```diff
@@ -1737,7 +1769,7 @@ def nlls_fit_tensor(design_matrix, data, weighting=None,
 
 def restore_fit_tensor(design_matrix, data, sigma=None, jac=True,
-                       return_S0_hat=False):
+                       return_S0_hat=False, fail_is_nan=False):
```
it should be exposed, since the API accepts a general list of keyword arguments, but you are right it should be added as a test to see if it works.

samcoveney replied:
tests added

samcoveney commented on dipy/reconst/dti.py:
```diff
@@ -1801,90 +1850,102 @@ def restore_fit_tensor(design_matrix, data, sigma=None, jac=True,
-                                                   flat_data[vox],
-                                                   'sigma',
-                                                   sigma))
+        try:  # FIXME: is this in correct place? if crashing on GMM, would we want to default to NL, rather than OLS??
```
okay, I have fixed this to use the original NL result (if we managed to calculate it), or the OLS result if the original NL failed (same checks as in the nls fitting function)

samcoveney replied:
I have left this alone for now. Perhaps we should default for NLLS sigma-weighted solution if GMM fails, provided the initial fit for NLLS sigma-weighted solution did not fail, but since the GMM usually fails due to bad sigma choice, it is not clear to me that the NLLS is a good default choice after all. Suggest to leave alone for now. The option `fail_is_nan` and new user warnings should help.

samcoveney commented on dipy/reconst/dti.py:
```diff
@@ -1801,90 +1850,102 @@ def restore_fit_tensor(design_matrix, data, sigma=None, jac=True,
     for vox in range(flat_data.shape[0]):
         if np.all(flat_data[vox] == 0):
-            raise ValueError("The data in this voxel contains only zeros")
+            raise ValueError("The data in this voxel contains only zeros") # FIXME: what to do if this is true? Crashing the code seems wrong
```
I will leave as it is, since the tensor API should fix it anyway (minimum value)

samcoveney commented:
I have finished the changes.

To summarize:
* The RESTORE algorithm now refits the LLS solution with clean data (outliers removed), before NLLS with clean data (whereas before, NLLS with clean data started optimization using LLS solution before outlier removal, which seemed wrong).
* I have added the option to return NaN values if NLLS fails to both `nlls_fit_tensor` and `restore_fit_tensor`, including adding tests for both functions. I have also caught errors for both functions that were previously not caught, requiring try and except to be in slightly different places than the original code
* I have modified the calculation of weights in `wls_fit_tensor` (same result) so that `_ols_fit_tensor` is not needed (although I have not removed this helper function)
* I have tidied up the initial OLS fit in both `nlls_fit_tensor` and `restore_fit_tensor`

I will rebase once this is all checked and approved.

samcoveney commented on dipy/reconst/dti.py:
```diff
@@ -1785,11 +1820,25 @@ def restore_fit_tensor(design_matrix, data, sigma=None, jac=True,
+        inv_design = np.linalg.pinv(design_matrix)
+        log_s = np.log(flat_data)
+        D = np.dot(inv_design, log_s.T).T
+    else:  # use ols_fit_tensor function directly, which means we make use of the iterator wrapper, and 'conditions' solution based on min_diffusivity
```
New benefit is tidier code, and again the iterator wrapper gets utilized

samcoveney commented:
Just added a new commit, so that the code can determine whether sigma varies with image/acquisition or varies with spatial position. I could think of no better way of doing this.

samcoveney commented:
Not heard back on this in a while, but I think I have responded to everything, so I will rebase this pull request soon

skoudoro commented on dipy/reconst/dti.py:
```diff
@@ -1378,12 +1402,23 @@ def wls_fit_tensor(design_matrix, data, return_S0_hat=False):
+    w = np.exp(fit_result @ design_matrix.T)
+
+    # the weighted problem design_matrix * w is much larger (differs per voxel)
+    if np.__version__ < '1.14':
```
can be dangerous, use `Version` from `packaging.version`

below an example :  

```python
In [1]: '1.4.0rc2' < '1.4.0'
Out[1]: False
# it should be True, release candidate lower
```

skoudoro reviewed (commented):
Any additional comments @arokem? 

Can we move forward with this PR.

thank you for this work @samcoveney!

arokem commented:
Sorry for the slowness here. I would really love to have a chance to run this before/after the changeset is applied, to make sure that it works as expected, but this is quite some work, so it's been slow going getting to it.

arokem commented:
OK - I ran the example we have with/without and this looks good to me. I'll go ahead and merge. 

Thanks @samcoveney !

samcoveney commented:
Thanks @arokem and @skoudoro , sorry I didn't get back to your about the version test/condition, hopefully we can easily fix this later if it causes any issues. Thanks again!

arokem commented:
Thank you! If you could make a quick PR with that fix, that would be really great!
